pm: Use fixed cpu id in single core case

nordic-krch opened the pull request:
When there are no multiple cores than fixing id to 0 saves few cycles.

Do not call ticks_expiring_sooner if not needed.

This small improvement saves few cycles but surprisingly significantly reduces `pm.c` code size 416 to 316 bytes in the sample that I've checked.

nordic-krch committed:
pm: Use fixed cpu id in single core case

When there are no multiple cores than fixing id to 0 saves few
cycles.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

JordanYates commented on subsys/pm/pm.c:
```diff
@@ -153,7 +154,8 @@ bool pm_system_suspend(int32_t kernel_ticks)
 	 */
 	events_ticks = pm_policy_next_event_ticks();
-	ticks = ticks_expiring_sooner(kernel_ticks, events_ticks);
+	ticks = (events_ticks != -1) ?
```
This seems unrelated to the other changes?

nordic-krch replied:
removed .

nashif commented on subsys/pm/pm.c:
```diff
@@ -49,6 +49,7 @@ static struct k_spinlock pm_notifier_lock;
  */
 static inline void pm_state_notify(bool entering_state)
 {
+	uint8_t id = (CONFIG_MP_MAX_NUM_CPUS == 1) ? 0 : _current_cpu->id;
```
s/id/cpu_id/

JordanYates commented:
Could ROM savings be obtained across other users in the codebase?
```
#if CONFIG_MP_MAX_NUM_CPUS == 1
#define CPU_ID 0
#else
#define CPU_ID _current_cpu->id
#endif
```

nordic-krch committed:
kernel_structs: Add macro CPU_ID for getting current CPU ID

In case of a single core it resolves to 0. In many cases it reduces
code size and execution time compared to reading id in runtime.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
Added `CPU_ID` in kernel_structs.h and used it in pm related code. I've briefly looked at generic code and found that in `kernel/timeslicing.c` there can be some small savings. There are places in `arch` where `_current_cpu->id` is used but i did not look into that.
I can post timeslicing PR once this one is merged to keep this PR focused on PM.

JordanYates commented on include/zephyr/kernel_structs.h:
```diff
@@ -257,6 +257,8 @@ __attribute_const__ struct k_thread *z_smp_current_get(void);
 #define _current _kernel.cpus[0].current
 #endif
 
+#define CPU_ID (CONFIG_MP_MAX_NUM_CPUS == 1) ? 0 : _current_cpu->id
```
Purely for the `NO_OPTIMIZATIONS`  case I think it would make sense to split the cases by `#if`.
If not, it should at least be surrounded by brackets.

bjarki-andreasen reviewed (approved):
Nice!

nordic-krch committed:
drivers: gpio: esp32: Add ESP32_ prefix to CPU_ID() macro

Driver is using local macro to get current CPU ID and it is now
in conflict with generic CPU_ID. Added prefix to avoid conflict.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>
